src/incidencehandler.cpp

```diff
@@ -367,37 +393,41 @@ KCalCore::Incidence::Ptr IncidenceHandler::incidenceToExport(KCalCore::Incidence
             // all-day event, so remove the DTEND before upsyncing.
-            LOG_DEBUG("Remove DTEND from" << incidence->uid());
+            LOG_DEBUG("Removing DTEND from" << incidence->uid());
             event->setDtEnd(KDateTime());
-        } else if (event->hasEndDate()) {
-            KDateTime dt = event->dtEnd();
-            // Event::dtEnd() is inclusive, but DTEND in iCalendar format is exclusive.
-            LOG_DEBUG("Adding +1 day to" << dt.toString() << "to make exclusive DTEND" << "for" << incidence->uid());
-            dt = dt.addDays(1);
-            event->setDtEnd(dt);
         } else {
-            KDateTime dt = event->dtStart().addDays(1);
-            LOG_DEBUG("Adding DTEND of DTSTART+1" << dt.toString() << "for" << incidence->uid());
+            KDateTime dt;
+            if (event->hasEndDate()) {
+                // Event::dtEnd() is inclusive, but DTEND in iCalendar format is exclusive.
+                dt = KDateTime(event->dtEnd().addDays(1).date(), event->dtEnd().timeSpec());
+                LOG_DEBUG("Adding +1 day to DTEND to make exclusive DTEND for" << incidence->uid() << ":" << dt.toString());
+            } else {
+                // No DTEND exists in event, but it's all day.  Set to DTSTART+1 to make exclusive DTEND.
+                dt = KDateTime(event->dtStart().addDays(1).date(), event->dtStart().timeSpec());
+                LOG_DEBUG("Setting DTEND to DTSTART+1 for" << incidence->uid() << ":" << dt.toString());
+            }
+            dt.setDateOnly(true);
+            LOG_DEBUG("Stripping time from date-only DTEND:" << dt.toString());
             event->setDtEnd(dt);
         }
     }
 
-    // if the time was added by us, remove it before upsyncing to the server
-    if (!event->customProperty("buteo", PROP_DTSTART_DATE_ONLY).isEmpty()) {
-        KDateTime dt = event->dtStart();
-        LOG_DEBUG("Strip time from start date" << dt.toString());
+    if (event->dtStart().isDateOnly()) {
+        KDateTime dt = KDateTime(event->dtStart().date(), event->dtStart().timeSpec());
         dt.setDateOnly(true);
         event->setDtStart(dt);
-        event->removeCustomProperty("buteo", PROP_DTSTART_DATE_ONLY);
+        LOG_DEBUG("Stripping time from date-only DTSTART:" << dt.toString());
     }
-    if (!event->customProperty("buteo", PROP_DTEND_DATE_ONLY).isEmpty()) {
-        KDateTime dt = event->dtEnd();
-        LOG_DEBUG("Strip time from end date" << dt.toString());
-        dt.setDateOnly(true);
-        event->setDtEnd(dt);
-        event->removeCustomProperty("buteo", PROP_DTEND_DATE_ONLY);
+
+    // setting dtStart/End changes the allDay value, so ensure it is still set to true if needed.
+    if (eventIsAllDay) {
+        event->setAllDay(true);
     }
 
+    // remove any markers that tell us that the time was added by us
+    event->removeCustomProperty("buteo", PROP_DTSTART_DATE_ONLY);
+    event->removeCustomProperty("buteo", PROP_DTEND_DATE_ONLY);
```

Comment:
Right, but what about the case where it was a locally-added event?  In that case, we probably should strip out time data, but the flags are not set (because they weren't synced).

Actually, why do those flags exist at all?  Why do we add time information to events synced from the remote server which are date-only?